### PR TITLE
NO REVIEW Distributed task manager for webadmin it

### PR DESCRIPTION
**Pull request by mbaechler**
This PR is for build status purpose only

### CODE REVIEW

**Comment by rouazana** [comment]
```
[6f53c417be2300360421b0674d4ed825c4ee1ed1] [ERROR] Failed to execute goal on project james-server-util: Could not resolve dependencies for project org.apache.james:james-server-util:jar:3.5.0-SNAPSHOT: Failed to collect dependencies at com.github.steveash.guavate:guavate:jar:1.0.0 -> com.google.guava:guava:jar:25.1-jre -> org.codehaus.mojo:animal-sniffer-annotations:jar:1.14: Failed to read artifact descriptor for org.codehaus.mojo:animal-sniffer-annotations:jar:1.14: Could not transfer artifact org.codehaus.mojo:animal-sniffer-parent:pom:1.14 from/to central (https://repo.maven.apache.org/maven2): /root/.m2/repository/org/codehaus/mojo/animal-sniffer-parent/1.14/animal-sniffer-parent-1.14.pom.part (No such file or directory) -> [Help 1]
```

**Comment by rouazana** [comment]
can you handle the migration to JUnit 5 at the same time please?

**Comment by mbaechler** [comment]
Is it still a "no review" PR?

**Comment by rouazana** [comment]
> Is it still a "no review" PR?

Yes, while the work is not done, we are breaking some webadmin functionnalities. I could remove the WIP prefix, but then we should only merge on a branch.

**Comment by rouazana** [comment]
replaced by https://github.com/linagora/james-project/pull/2746

File: server/container/guice/cassandra-rabbitmq-guice/src/main/java/org/apache/james/modules/TaskSerializationModule.java [hunk 1]
```diff
@@ -56,6 +60,11 @@ protected void configure() {
         return FullReindexingTask.module(performer);
     }
 
+    @ProvidesIntoSet
+    public TaskDTOModule<?, ?> deleteMailsFromMailQueueTask(MailQueueFactory<?> mailQueueFactory) {
+        return DeleteMailsFromMailQueueTaskDTO.module((MailQueueFactory<ManageableMailQueue>) mailQueueFactory);
```
Comment (mbaechler) [inline server/container/guice/cassandra-rabbitmq-guice/src/main/java/org/apache/james/modules/TaskSerializationModule.java]: what makes you think this assumption stands?

File: server/protocols/webadmin-integration-test/src/test/java/org/apache/james/webadmin/integration/WebAdminServerIntegrationTest.java [hunk 2]
```diff
@@ -454,4 +455,28 @@ public void deleteMailsTasksShouldCompleteWhenSenderIsValid() {
                 .body("status", is("completed"));
     }
 
+    @Test
+    public void reprocessingAllTaskShouldCreateATask() throws Exception {
```
Comment (mbaechler) [inline server/protocols/webadmin-integration-test/src/test/java/org/apache/james/webadmin/integration/WebAdminServerIntegrationTest.java]: does it test that the task has been serialized?

File: server/protocols/webadmin-integration-test/src/test/java/org/apache/james/webadmin/integration/WebAdminServerIntegrationTest.java [hunk 3]
```diff
@@ -595,4 +596,23 @@ public void deletedMessageVaultRestoreShouldCreateATask() throws Exception {
         .then()
             .body("status", is("completed"));
     }
+
+    @Test
+    public void deletedMessageVaultExportShouldCreateATask() throws Exception {
```
Comment (mbaechler) [inline server/protocols/webadmin-integration-test/src/test/java/org/apache/james/webadmin/integration/WebAdminServerIntegrationTest.java]: does it test the task serialization?

File: server/protocols/webadmin-integration-test/src/test/java/org/apache/james/webadmin/integration/WebAdminServerIntegrationTest.java [hunk 4]
```diff
@@ -498,6 +504,29 @@ public void reprocessingAllTaskShouldCreateATask() throws Exception {
                 .body("taskId", is(Matchers.notNullValue()));
     }
 
+    @Test
+    public void reprocessingOneTaskShouldCreateATask() throws Exception {
```
Comment (mbaechler) [inline server/protocols/webadmin-integration-test/src/test/java/org/apache/james/webadmin/integration/WebAdminServerIntegrationTest.java]: does it test that the task is serialized?

File: server/protocols/webadmin-integration-test/src/test/java/org/apache/james/webadmin/integration/WebAdminServerIntegrationTest.java [hunk 5]
```diff
@@ -678,4 +678,15 @@ public void fixingReIndexingShouldNotFailWhenNoMail() {
             .body("type", is("ErrorRecoveryIndexation"));
     }
 
+    @Test
+    public void postRedeliverAllEventsShouldCreateATask() {
```
Comment (mbaechler) [inline server/protocols/webadmin-integration-test/src/test/java/org/apache/james/webadmin/integration/WebAdminServerIntegrationTest.java]: does it test that the task is serialized?

File: server/protocols/webadmin-integration-test/src/test/java/org/apache/james/webadmin/integration/WebAdminServerIntegrationTest.java [hunk 6]
```diff
@@ -688,5 +697,34 @@ public void postRedeliverAllEventsShouldCreateATask() {
...
+            .eventId(Event.EventId.of(uuid))
+            .user(User.fromUsername(USERNAME))
+            .sessionId(MailboxSession.SessionId.of(452))
+            .mailboxId(InMemoryId.of(453))
+            .mailboxPath(MailboxPath.forUser(USERNAME, "Important-mailbox"))
+            .build();
+
+        guiceJamesServer
+            .getInjector()
+            .getInstance(EventDeadLetters.class)
```
Comment (mbaechler) [inline server/protocols/webadmin-integration-test/src/test/java/org/apache/james/webadmin/integration/WebAdminServerIntegrationTest.java]: why don't you publish it as a probe?

File: server/protocols/webadmin-integration-test/src/test/java/org/apache/james/webadmin/integration/WebAdminServerIntegrationTest.java [hunk 7]
```diff
@@ -688,5 +697,34 @@ public void postRedeliverAllEventsShouldCreateATask() {
...
+            .store(group, event, insertionId)
+            .block();
 
+        given()
+            .queryParam("action", "reDeliver")
+            .when()
+            .post("/events/deadLetter/groups/" + group.asString())
+            .then()
+            .statusCode(HttpStatus.CREATED_201)
```
Comment (mbaechler) [inline server/protocols/webadmin-integration-test/src/test/java/org/apache/james/webadmin/integration/WebAdminServerIntegrationTest.java]: does it test that the task is serialized?

File: server/protocols/webadmin-integration-test/src/test/java/org/apache/james/webadmin/integration/WebAdminServerIntegrationTest.java [hunk 8]
```diff
@@ -727,4 +727,35 @@ public void postRedeliverGroupEventsShouldCreateATask() {
...
+            .getInstance(EventDeadLetters.class)
+            .store(group, event, insertionId)
+            .block();
+
+        given()
+            .queryParam("action", "reDeliver")
+            .when()
+            .post("/events/deadLetter/groups/" + group.asString() + "/" + insertionUuid)
+            .then()
+            .statusCode(HttpStatus.CREATED_201)
```
Comment (mbaechler) [inline server/protocols/webadmin-integration-test/src/test/java/org/apache/james/webadmin/integration/WebAdminServerIntegrationTest.java]: does it test that the task is serialized?

File: server/container/guice/cassandra-rabbitmq-guice/src/main/java/org/apache/james/modules/TaskSerializationModule.java [hunk 9]
```diff
@@ -110,6 +110,11 @@ protected void configure() {
         return CassandraMappingsSolveInconsistenciesTask.module(migration, dao);
     }
 
+    @ProvidesIntoSet
+    public TaskDTOModule<?, ?> clearMailQueueTask(MailQueueFactory<?> mailQueueFactory) {
+        return ClearMailQueueTaskDTO.module((MailQueueFactory<ManageableMailQueue>) mailQueueFactory);
```
Comment (mbaechler) [inline server/container/guice/cassandra-rabbitmq-guice/src/main/java/org/apache/james/modules/TaskSerializationModule.java]: what make you think this assumption will stand?

File: server/container/guice/cassandra-rabbitmq-guice/src/main/java/org/apache/james/modules/TaskSerializationModule.java (hunk 1 above)
Comment (rouazana) [reply server/container/guice/cassandra-rabbitmq-guice/src/main/java/org/apache/james/modules/TaskSerializationModule.java]: all the injection code is based on this assumption :/

File: server/container/guice/cassandra-rabbitmq-guice/src/main/java/org/apache/james/modules/TaskSerializationModule.java (hunk 9 above)
Comment (rouazana) [reply server/container/guice/cassandra-rabbitmq-guice/src/main/java/org/apache/james/modules/TaskSerializationModule.java]: all the injection code is based on this assumption :/

File: server/protocols/webadmin-integration-test/src/test/java/org/apache/james/webadmin/integration/WebAdminServerIntegrationTest.java (hunk 2 above)
Comment (rouazana) [reply server/protocols/webadmin-integration-test/src/test/java/org/apache/james/webadmin/integration/WebAdminServerIntegrationTest.java]: yes, but anyway I will improve the tests to be more coherent

File: server/protocols/webadmin-integration-test/src/test/java/org/apache/james/webadmin/integration/WebAdminServerIntegrationTest.java (hunk 3 above)
Comment (rouazana) [reply server/protocols/webadmin-integration-test/src/test/java/org/apache/james/webadmin/integration/WebAdminServerIntegrationTest.java]: yes, but anyway I will improve the tests to be more coherent

File: server/protocols/webadmin-integration-test/src/test/java/org/apache/james/webadmin/integration/WebAdminServerIntegrationTest.java (hunk 4 above)
Comment (rouazana) [reply server/protocols/webadmin-integration-test/src/test/java/org/apache/james/webadmin/integration/WebAdminServerIntegrationTest.java]: yes, but anyway I will improve the tests to be more coherent

File: server/protocols/webadmin-integration-test/src/test/java/org/apache/james/webadmin/integration/WebAdminServerIntegrationTest.java (hunk 5 above)
Comment (rouazana) [reply server/protocols/webadmin-integration-test/src/test/java/org/apache/james/webadmin/integration/WebAdminServerIntegrationTest.java]: yes, but anyway I will improve the tests to be more coherent

File: server/protocols/webadmin-integration-test/src/test/java/org/apache/james/webadmin/integration/WebAdminServerIntegrationTest.java (hunk 6 above)
Comment (rouazana) [reply server/protocols/webadmin-integration-test/src/test/java/org/apache/james/webadmin/integration/WebAdminServerIntegrationTest.java]: what value does it add?

File: server/protocols/webadmin-integration-test/src/test/java/org/apache/james/webadmin/integration/WebAdminServerIntegrationTest.java (hunk 7 above)
Comment (rouazana) [reply server/protocols/webadmin-integration-test/src/test/java/org/apache/james/webadmin/integration/WebAdminServerIntegrationTest.java]: yes, but anyway I will improve the tests to be more coherent

File: server/protocols/webadmin-integration-test/src/test/java/org/apache/james/webadmin/integration/WebAdminServerIntegrationTest.java (hunk 8 above)
Comment (rouazana) [reply server/protocols/webadmin-integration-test/src/test/java/org/apache/james/webadmin/integration/WebAdminServerIntegrationTest.java]: yes, but anyway I will improve the tests to be more coherent

File: server/container/guice/cassandra-rabbitmq-guice/src/main/java/org/apache/james/modules/TaskSerializationModule.java (hunk 1 above)
Comment (mbaechler) [reply server/container/guice/cassandra-rabbitmq-guice/src/main/java/org/apache/james/modules/TaskSerializationModule.java]: I don't understand why we can't ask for the actual type in the method signature

File: server/protocols/webadmin-integration-test/src/test/java/org/apache/james/webadmin/integration/WebAdminServerIntegrationTest.java (hunk 6 above)
Comment (mbaechler) [reply server/protocols/webadmin-integration-test/src/test/java/org/apache/james/webadmin/integration/WebAdminServerIntegrationTest.java]: "rule of least power"
Giving access to every single classes of james breaks the encapsulation goal of an object oriented design.
As soon as you offer such an API, it will be abused and it will becomes impossible to go forward and make changes to the code because some code will make assumptions about the objects it can have access to and how it behaves.
It's "API design" vs "access everything from everywhere"
In short, publishing the Injector is Evil

Comment (rouazana) [reply server/protocols/webadmin-integration-test/src/test/java/org/apache/james/webadmin/integration/WebAdminServerIntegrationTest.java]: >  Giving access to every single classes of james breaks the encapsulation goal of an object oriented design.

The access is already granted into any production code through the use of `@Inject` annotation.
It just add access to test code, which can greatly simplify writing some tests. So finally I still don't see the evil part here.

File: server/container/guice/cassandra-rabbitmq-guice/src/main/java/org/apache/james/modules/TaskSerializationModule.java (hunk 9 above)
Comment (rouazana) [reply server/container/guice/cassandra-rabbitmq-guice/src/main/java/org/apache/james/modules/TaskSerializationModule.java]: see for example https://github.com/linagora/james-project/blob/master/server/protocols/webadmin/webadmin-mailqueue/src/main/java/org/apache/james/webadmin/routes/MailQueueRoutes.java#L102

File: server/protocols/webadmin-integration-test/src/test/java/org/apache/james/webadmin/integration/WebAdminServerIntegrationTest.java (hunk 6 above)
Comment (chibenwa) [reply server/protocols/webadmin-integration-test/src/test/java/org/apache/james/webadmin/integration/WebAdminServerIntegrationTest.java]: Developping probes is a bit verbose but not hard. I don't see why we should refrain nor change the decision for conveniance on that point.

File: server/container/guice/cassandra-rabbitmq-guice/src/main/java/org/apache/james/modules/TaskSerializationModule.java (hunk 9 above)
Comment (chibenwa) [reply server/container/guice/cassandra-rabbitmq-guice/src/main/java/org/apache/james/modules/TaskSerializationModule.java]: This sucks.

One of the possible proposition would be not to propose Manageable part for a James-distribted MDA product, and I believe even if we do not do that move being flexible on this is a good thing for users assembling their own Guice products.

I'd rather create a separate issue about that 'the mailqueue is manageable' assumption. Maybe we need a clear API split?

File: server/protocols/webadmin-integration-test/src/test/java/org/apache/james/webadmin/integration/WebAdminServerIntegrationTest.java (hunk 6 above)
Comment (rouazana) [reply server/protocols/webadmin-integration-test/src/test/java/org/apache/james/webadmin/integration/WebAdminServerIntegrationTest.java]: I did not say it's hard. And yes verbosity is an inconvenient. I just would like to understand the ins and outs of this decision.

Comment (chibenwa) [reply server/protocols/webadmin-integration-test/src/test/java/org/apache/james/webadmin/integration/WebAdminServerIntegrationTest.java]: In `GuiceJamesServer.java` , `public Injector getInjector() {...}` is public.

This is a major drawback.

Think of the powers we give to people embedding a James guice in another application.

Comment (rouazana) [reply server/protocols/webadmin-integration-test/src/test/java/org/apache/james/webadmin/integration/WebAdminServerIntegrationTest.java]: see https://github.com/linagora/james-project/pull/2748/commits/0ec7966eecbd49afd55599fbcc510cafaf806b77

File: server/container/guice/cassandra-rabbitmq-guice/src/main/java/org/apache/james/modules/TaskSerializationModule.java (hunk 1 above)
Comment (rouazana) [reply server/container/guice/cassandra-rabbitmq-guice/src/main/java/org/apache/james/modules/TaskSerializationModule.java]: see https://issues.apache.org/jira/browse/JAMES-2926

File: server/container/guice/cassandra-rabbitmq-guice/src/main/java/org/apache/james/modules/TaskSerializationModule.java (hunk 9 above)
Comment (rouazana) [reply server/container/guice/cassandra-rabbitmq-guice/src/main/java/org/apache/james/modules/TaskSerializationModule.java]: see https://issues.apache.org/jira/browse/JAMES-2926
